### ISSUE TITLE
Change hash scheme to make validation task more efficient

### ISSUE THREAD

**Issue by dchiquito** [dandi/dandi-api#145]
An initial proposal:

A new model `ChunkValidation` is added to the DB that tracks each chunk of the blob validation.

When the `validate` endpoint is called:
* The server will divide the blob into a number of distinct chunks, let's say 100 MB (this will be tuned during development to balance task run time against task count). For each chunk, a new `ChunkValidation` is created that tracks the blob being validated, the beginning and end range of the chunk, the current validation status (initially `IN_PROGRESS`), and the final result of the checksum calculation (initially `null`).
* A new `validate_chunk` celery task is queued for each `ChunkValidation`. It will stream the bytes for the chunk and calculate the sha256 of the chunk, then store that sha256 in the `ChunkValidation` and update the status. If the task fails, it is retried 5 times (for example). All these tasks can be run concurrently.
* After each `validate_chunk` task finishes, it checks to see if all the `ChunkValidations` have completed. If so, it queues a `collate_checksum` celery task.
* The `collate_checksum` task concatenates all the checksums in the `ChunkValidations` in order, then calculates the sha256 of the resulting string. This is the final checksum for the blob.

The CLI would have to implement basically the same thing, but without the database or celery tasks.

Some points to consider:
* To bastardize some notation, the underlying calculation is `sha256(str(sha256(chunk[0])) + str(sha256(chunk[1])) + ...)`. sha256 may not be the best hashing algorithm, I assume there are faster algorithms. Simple concatenation may not be the best aggregator, although I don't see any security vulnerabilities.
* The CLI would basically have to reimplement the same code that the API does.
* I found a few python libraries that implement Merkle trees, but this implementation is so simplistic that involving a library would probably be more work than implementing it ourselves.
* Chunking out the calculation makes it trivial to add a progress counter, which is a nice UX feature.

**Comment by yarikoptic** [comment]
I would just do exactly what ipfs does. Then we could hypothetically in the future (an)use that checksum also to query ipfs for the content .

**Comment by yarikoptic** [comment]
FWIW, https://pascalprecht.github.io/posts/content-identifiers-in-ipfs is an introductory blog post which shines the light on details behind default ipfs approach... The core principles nicely summarized in https://stackoverflow.com/a/40999931/1265472 as:

2

A file in IPFS is 'transformed' into a Unixfs file, which is a representation of files in a DAG, in your example, you are hashing directly your multihash.txt with sha2-256, but what happens inside IPFS is:

- file gets chunked into 256KiB pieces
- each chunk goes into a DAG node inside a Unixfs protobuf https://github.com/ipfs/js-ipfs-unixfs
- a dag is created with links to all the chunks.

Even though 256KiB sounds like too little for our use case, I think it would be just fine since we can split into large blocks (per individual worker) ... we just need to find an implementation for the `ipfs add --only-hash` invocation and such that we could distribute computation over multiple workers

**Comment by satra** [comment]
i don't know why s3 doesn't do this but glacier does this: https://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html

**Comment by mgrauer** [comment]
I vote for either @dchiquito 's first proposal or else @satra 's treehash idea.

@yarikoptic 's approach sounds more complicated, but I admit I haven't fully understood it. Is there python library support for this approach and can it be parallelized/chunked?

I believe:

* the treehash idea could work with the original proposal (which is a map-reduce)
* with only the change of how `collate_checksum` works
* and that continues to provide all of the benefits laid out in the proposal

**Comment by yarikoptic** [comment]
> @yarikoptic 's approach sounds more complicated, but I admit I haven't fully understood it. Is there python library support for this approach and can it be parallelized/chunked?

indeed unfortunately it is indeed more complicated.  I am yet to check for the library or how it possibly relate to treehash (I would expect it to do something similar but just + extra ipfs specific metadata on type of the load etc).  For whatever choice we make, I think having an existing implementation (instead of crafting our own) should pretty much be a requirement.

**Comment by satra** [comment]
here is an attempt to separate out a few pieces. the worker can of course do range requests for bytes on s3. 

<details> <summary> An untested treehash </summary>

```python
import asyncio
import math
from hashlib import sha256
import os


async def worker(file, offset, size, hashfunc):
    with open(file, 'rb') as fp:
        fp.seek(offset)
        return hashfunc(fp.read(size)).digest()


async def hashtree(offset, size, file, chunk, hashfunc):
    if size <= chunk:
        return await worker(file, offset, size, hashfunc)
    N = math.ceil(size/chunk)
    if N % 2 == 1:
        mid = int(math.ceil(N/2)) * chunk
    else:
        mid = int(N/2) * chunk
    value = await asyncio.gather(
        hashtree(offset, mid, file, chunk, hashfunc),
        hashtree(mid, size - mid, file, chunk, hashfunc))
    value = hashfunc(value[0] + value[1]).digest()
    return value


if __name__ == "__main__":
    chunk = int(1024*1024)
    file = "10g.bin"
    size = os.stat(file).st_size
    print(size)
    print(math.ceil(size/chunk))

    out = asyncio.run(hashtree(0, size, file, chunk, sha256))
    print(out)
```
</details>

**Comment by satra** [comment]
just a note that this is not quite the merkle tree, since i split from top down, but it could be rewritten quite easily with precalculated offsets and branching from the bottom up. 

my main reason was to consider what we could do both locally and on the server that allowed for large files and what that meant for depth and number of leaves. for example, at the AWS max s3 object size of 5TiB, the number of leave nodes would be about a million if we use 5MiB chunks. and the file that i uploaded was about 3.3. i might strongly suggest that the worker mount the bucket using Amazon FSx when computing the validation check for any files larger than some size empirically determined.

**Comment by yarikoptic** [comment]
Note: for client we will need serial/streaming implementation (doable) version so we could compute it along with simple sha256 we have now in a single read pass

**Comment by mgrauer** [comment]
I've thought about this more. What do we get from the treehash that is an improvement over the list hash calculation? The list hash is simpler, which is an advantage.

**Comment by satra** [comment]
there are potential benefits in the short and long terms.

1. if you are caching the hash, you don't have to store the entire list, which at 256K for 5TB is a long list.
2. if you ever wanted to verify which sequences were broken the treehash allows you to do so more efficiently (if you want more efficiency you can see hashgraph)
3. if we do the same implementation as glacier, it would allow us to move any chunk to glacier quite easily. 

none of these are perhaps migration blockers, but given how easy it is to compute tree hash, why wouldn't you? the serial hash would require you to store the entire list of hashes in a database, you could be much for efficient with a serial tree hash (not the above implementation, but the one yarik mentioned - also easy to code and parallelize). you don't store each chunk in a database, you store the chunk at a given depth, which accounts for potentially 100s of data chunks thus minimize revalidation if the worker crashes. 

personally i'm more worried why the celery worker crashes. the tree hash algorithm should not require any db storage because network interruptions can be easily handled be handled at the hash of bytestream worker level.

**Comment by satra** [comment]
@dchiquito - a question about your suggested implementation. why are you thinking a database model is required here, over a resilient worker that can withstand network interruptions?

**Comment by dchiquito** [comment]
Building a resilient worker is something of an unachievable goal. We can sink as much time as you like into finding and addressing all the common failure modes, but we'll never think of everything, and every edge case means more code to write and maintain. 
Alternatively, we simply make the tasks idempotent. It's easy to do, and retrying will resolve any transient issues much more reliably than trying to catch them all explicitly.

Also, we need a way to store the first checksums while the rest are still being calculated. This is the simplest async way to go about doing that.

**Comment by yarikoptic** [comment]
I believe this one was superseded by now closed #146
